OpenRA.Mods.D2k/Traits/SpiceBloom.cs

```diff
@@ -91,45 +103,50 @@ public void Tick(Actor self)
 
 		public void Killed(Actor self, AttackInfo e)
 		{
-			var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
-			if (pieces < info.Pieces[0])
-				pieces = info.Pieces[0];
-
-			for (var i = 0; pieces > i; i++)
+			if (!string.IsNullOrEmpty(info.Weapon))
 			{
-				var cells = OpenRA.Traits.Util.RandomWalk(self.Location, self.World.SharedRandom);
-				var cell = cells.Take(info.Range).SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p)).Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
-				if (cell == null)
-					cell = cells.Take(info.Range).Random(self.World.SharedRandom);
+				var pieces = self.World.SharedRandom.Next(info.Pieces[0], info.Pieces[1]) * ticks / growTicks;
+				if (pieces < info.Pieces[0])
+					pieces = info.Pieces[0];
 
-				var args = new ProjectileArgs
+				var cells = self.World.Map.FindTilesInAnnulus(self.Location, 1, info.Range)
+					.SkipWhile(p => resLayer.GetResource(p) == resType && resLayer.IsFull(p));
+
+				for (var i = 0; pieces > i; i++)
 				{
-					Weapon = self.World.Map.Rules.Weapons[info.Weapon.ToLowerInvariant()],
-					Facing = 0,
+					var cell = cells.Cast<CPos?>().RandomOrDefault(self.World.SharedRandom);
+					if (cell == null)
+						cell = cells.Random(self.World.SharedRandom);
```

Comment:
Tbh I didn't (and still don't) fully understand this piece of code from #8987, so I just avoided touching it too much.